core/templates/services/playthrough.service.ts

```diff
@@ -1,334 +1,359 @@
+import { Playthrough, PlaythroughObjectFactory } from
+  'domain/statistics/PlaythroughObjectFactory';
+import { PlaythroughBackendApiService } from
+  'domain/statistics/playthrough-backend-api.service';
+import { ServicesConstants } from 'services/services.constants';
+import { Stopwatch } from 'domain/utilities/stopwatch.model';
+
+class CyclicStateTransitionsTracker {
+  /** A path of visited states without any repeats. */
+  private pathOfVisitedStates: string[];
+  /** The most recently discovered cycle of visited states. */
+  private cycleOfVisitedStates: string[];
+  private numLoops: number;
+
+  constructor(initStateName: string) {
+    this.pathOfVisitedStates = [initStateName];
+    this.cycleOfVisitedStates = [];
+    this.numLoops = 0;
+  }
+
+  foundAnIssue(): boolean {
+    return this.numLoops >= ServicesConstants.NUM_REPEATED_CYCLES_THRESHOLD;
+  }
+
+  /**
+    * Records learner's transition to a new state into this tracker's path of
+    * visited states.
+    *
+    * If appending the new state would introduce a duplicate state name, then a
+    * cycle has been found. Specifically, if pushing the newest state (N) onto
+    * the path would result in the following pattern:
+    *
+    *    [ ... , N , ... , N ]
+    *
+    * then we update this tracker's latest cycle discovery. The cycle is defined
+    * as the current path of visited states with all states prior to the first
+    * occurrence of N discarded:
+    *
+    *    [ ... // N , ... , N ]  =>  [ N , ... , N ]
+    *
+    * If this *exact* cycle has been discovered before (NOTE: rotations of a
+    * cycle are considered to be different from each other),then we increase the
+    * tracked number of cycle occurrences.
+    * Otherwise, the tracker is completely reset to 1.
+    *
+    * Finally, the path of visited states is reset to a value of [ N ], in hopes
+    * that the exact same cycle is discovered enough times to be considered an
+    * issue.
+    */
+  recordStateTransition(destStateName: string): void {
+    if (this.currStateName() === destStateName) {
+      return;
+    }
+    if (this.pathOfVisitedStates.includes(destStateName)) {
+      const cycleOfVisitedStates = (
+        this.makeCycle(this.pathOfVisitedStates.indexOf(destStateName)));
+      if (angular.equals(this.cycleOfVisitedStates, cycleOfVisitedStates)) {
+        this.numLoops += 1;
+      } else {
+        this.cycleOfVisitedStates = cycleOfVisitedStates;
+        this.numLoops = 1;
+      }
+       this.pathOfVisitedStates.length = 0;
+    }
+    this.pathOfVisitedStates.push(destStateName);
+  }
+
+  generateIssueCustomizationArgs(): CyclicStateTransitionsCustomizationArgs {
+    return {
+      state_names: {value: this.cycleOfVisitedStates}
+    };
+  }
+
+  private makeCycle(collisionIndex: number): string[] {
+    const collision = this.pathOfVisitedStates[collisionIndex];
+    const cycleWithNoCollision =
+    this.pathOfVisitedStates.slice(collisionIndex);

```

Comment:
```suggestion
    const cycleWithNoCollision = this.pathOfVisitedStates.slice(collisionIndex);
```

Reply:
It throws eslint error that line has max length of 80 but now it is 81.

Reply:
```suggestion
    const cycleWithNoCollision = (
      this.pathOfVisitedStates.slice(collisionIndex));

```